web/pandas/pdeps/0013-logical-type-system.md

```diff
@@ -0,0 +1,161 @@
+
+## Proposal
+
+Derived from the hierarchical visual in the previous section, this PDEP proposes that pandas supports at least all of the following _logical_ types, excluding any type widths for brevity:
+
+  - Signed Integer
+  - Unsigned Integer
+  - Floating Point
+  - Fixed Point
+  - Boolean
+  - Date
+  - Datetime
+  - Duration
+  - Interval
+  - Period
+  - Binary
+  - String
+  - Dictionary
+  - List
+  - Struct
+
+To ensure we maintain all of the current functionality of our existing type system(s), a base type structure would need to look something like:
+
+```python
+class BaseType:
+
+    @property
+    def dtype_backend -> Literal["pandas", "numpy", "pyarrow"]:
+        """
+        Library is responsible for the array implementation
+        """
+        ...
+
+    @property
+    def physical_type:
+        """
+        How does the backend physically implement this logical type? i.e. our
+        logical type may be a "string" and we are using pyarrow underneath -
+        is it a pa.string(), pa.large_string(), pa.string_view() or something else?
+        """
+        ...
+
+    @property
+    def missing_value_marker -> pd.NA|np.nan:
+        """
+        Sentinel used to denote missing values
+        """
+        ...
+```
+
+The theory behind this PDEP is that most users _should not_ care about the physical type that is being used. But if the abstraction our logical type provides is too much, a user could at least inspect and potentially configure which physical type to use.
+
+With regards to how we may expose such types to end users, there are two currently recognized proposals. The first would use factory functions to create the logical type, i.e. something like:
+
+```python
+pd.Series(["foo", "bar", "baz"], dtype=pd.string()  # assumed common case
+pd.Series(["foo", "bar", "baz"], dtype=pd.string(missing_value_marker=np.nan)
+pd.Series(["foo", "bar", "baz"], dtype=pd.string(physical_type=pa.string_view())
+```
+
+Another approach would be to classes:
+
+```python
+pd.Series(["foo", "bar", "baz"], dtype=pd.StringDtype()
+pd.Series(["foo", "bar", "baz"], dtype=pd.StringDtype(missing_value_marker=np.nan)
+pd.Series(["foo", "bar", "baz"], dtype=pd.StringDtype(physical_type=pa.string_view())
+```
+Note that the class-based approach would reuse existing classes like ``pd.StringDtype`` but change their purpose, whereas the factory function would more explicitly be a new approach. This is an area that requires more discussion amongst the team.
+
+## String Type Arguments
+
+This PDEP proposes that we maintain only a small subset of string arguments that can be used to construct logical types. Those string arguments are:
```

Comment:
off the top of my head, this would mean deprecating M8[unit], m8[unit], timedelta64[unit], period[freq], interval, interval[subtype, closed], category.  (the first three of which are supported by `np.dtype(...)`).  Just checking that these are all intentional.

Reply:
I definitely would want to deprecate using the NumPy types since those would be circumventing the logical type system. I am -0.5 on continuing to allow timedelta64 because that also refers directly to the extension array implementation name rather than a logical type. 

Period seems logical to continue supporting. I don't think I've ever constructed an interval that way before so I'll withhold an opinion on that for now, but generally these are offered for legacy support and not the ideal way of creating types going forward

Reply:
> I definitely would want to deprecate using the NumPy types

See also some of the discussion in https://github.com/pandas-dev/pandas/issues/58243.

But IMO it's a very hard step to actually deprecate specifying numpy dtypes, given how widespread this usage is. And certainly for those types where there is an unambiguous mapping to our new logical types (like np.int64), I am not fully sure what the benefit is about deprecating that. 
In any case, if we would deprecate that, it would need to go very slowly, and we should only do that a while _after_ introducing the new pandas types.

Reply:
> And certainly for those types where there is an unambiguous mapping to our new logical types (like np.int64), I am not fully sure what the benefit is about deprecating that.

The benefit would be having some consistency in the type constructors we allow and it would decouple the user focus on something have to be NumPy.

I agree it would be a very long deprecation, but on the flip side I don't think a user really believes they are subscribing into any particular set of features when they ask for a `np.int64` data type. They really just want a 64 bit integer. 

Especially with integer types being zero copy there is very little benefit to asking for a `np.int64`, and I would argue that is even counterproductive with the possibility of missing values being introduced

Reply:
I'm leaning towards thinking deprecating these would cause a lot of churn for users without much clear benefit.  To the extent that you want to encourage users to use different patterns, I'm fine with that.

Reply:
Allowing users to construct physical types like `np.int64` alongside logical types like `pd.int64` I think defeats the purpose of adding logical types though; in such a case we are just adding more type construction options, which this PDEP is trying to solve

There also is a clear benefit to `pd.int64` over `np.int64`, namely the former has flexibility to better handle missing values whereas the latter locks you into numpy semantics

Reply:
> Allowing users to construct physical types like np.int64 alongside logical types like pd.int64 I think defeats the purpose of adding logical types though; 

In my mind, we wouldn't allow them to create a Series with a "physical type", we would just accept `np.int64` as input for the `dtype` argument, and specifying `np.int64` or `pd.int64()` would give the exact same result: a Series with whatever is the default int64 dtype at that point (right now that is the numpy based int64, in the future maybe a numpy masked int64 or pyarrow int64).

It's certainly true that this doesn't help with the "too many ways" to specify a dtype, and this PDEP would add yet another way, but at least a way that is consistent across all dtypes, and which ensures there is a clear mapping of whathever random input to the resulting dtype.

For example, `np.int64` is actually not even a dtype but a scalar constructor (`np.dtype("int64")` is the dtype), but conveniently accepted as a dtype argument in numpy. Just like the python types are also accepted, in addition to strings. 
So with the logical types, you can say that (for now) all of `np.int64`, `np.dtype("int64")`, `"int64"`, `int`, `"int"` all map to the same as `pd.int64()`, and longer term we could discuss reducing the number of aliases being accepted.

(I didn't do that with this comment .. ;), but personally I think we should maybe better focus first on the actual (logical) dtypes (the new part in this PDEP), and we can still later consider which aliases should be allowed or not. Short term we need to keep them all working anyway)

Reply:
I think that's fair. Providing compat is going to be a fact of life for this. Ultimately I don't think it would be that hard for a user to swap out `np.int64` for the logical constructor (in such an instance it would just be find/replace), but yea I agree that can be something cleaned up over time. Its not critical to execution of this PDEP